pkg/cloudprovider/providers/vsphere/nodeuuidcache.go

```diff
@@ -0,0 +1,184 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vsphere
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/golang/glog"
+	"k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+)
+
+const (
+	// Since node name can be of length 253 and uuid is of length 36
+	// given 1MB data limit of configmap, we can allow upto 3600 entries.
+	maxDataEntries = 3600
+)
+
+type nodeUUID struct {
+	uuid    string
+	deleted bool
+}
+
+type nodeUUIDCache struct {
+	nodeUUIDMap   map[string]*nodeUUID
+	kubeClient    clientset.Interface
+	configMapName string
+	namespace     string
+	cacheLock     sync.RWMutex
+}
+
+func newNodeUUIDCache(configMapName, namespace string) *nodeUUIDCache {
+	cache := &nodeUUIDCache{configMapName: configMapName, namespace: namespace}
+	cache.nodeUUIDMap = map[string]*nodeUUID{}
+	return cache
+}
+
+func (cache *nodeUUIDCache) addNode(nodeName, uuid string) error {
+	cache.cacheLock.Lock()
+	defer cache.cacheLock.Unlock()
+	configMap, err := cache.loadFromConfigMap()
+	if err != nil {
+		return fmt.Errorf("error loading uuid in configmap %s with %v", cache.configMapName, err)
+	}
+	cache.nodeUUIDMap[nodeName] = &nodeUUID{uuid: uuid}
+	configMap.Data = cache.getUUIData()
```

Comment:
I would suggest, to persist information that nodes that are marked as deleted in configmap. Since if kube-controller-manager restarts, vSphere Cloud Provider can keep the nodes which are marked as deleted if there is a need to prune the configmap.